File: WCPMVCVB/Controllers/DemoPrintCommandsController.vb
Imports System.Web.Mvc

Imports Neodynamic.SDK.Web

Namespace Controllers
    Public Class DemoPrintCommandsController
        Inherits Controller

        ' GET: DemoPrintCommands
        Function Index() As ActionResult

            ViewData("WCPScript") = Neodynamic.SDK.Web.WebClientPrint.CreateScript(Url.Action("ProcessRequest", "WebClientPrintAPI", Nothing, HttpContext.Request.Url.Scheme), Url.Action("PrintCommands", "DemoPrintCommands", Nothing, HttpContext.Request.Url.Scheme), HttpContext.Session.SessionID)

            Return View()
        End Function

        'sid: user session id who is requesting a ClientPrintJob
        <AllowAnonymous>
        Public Sub PrintCommands(sid As String)
            If WebClientPrint.ProcessPrintJob(System.Web.HttpContext.Current.Request.Url.Query) Then
                Dim app As HttpApplicationStateBase = HttpContext.Application

                'Create a ClientPrintJob obj that will be processed at the client side by the WCPP
                Dim cpj As New ClientPrintJob()

                'get printer commands for this user id
                Dim printerCommands As Object = app(sid + PRINTER_COMMANDS)
                If printerCommands IsNot Nothing Then
                    cpj.PrinterCommands = printerCommands.ToString()
                    cpj.FormatHexValues = True
                End If

                'get printer settings for this user id
                Dim printerTypeId As Integer = CInt(app(sid + PRINTER_ID))

                If printerTypeId = 0 Then
                    'use default printer
                    cpj.ClientPrinter = New DefaultPrinter()
                ElseIf printerTypeId = 1 Then
                    'show print dialog
                    cpj.ClientPrinter = New UserSelectedPrinter()
                ElseIf printerTypeId = 2 Then
                    'use specified installed printer
                    cpj.ClientPrinter = New InstalledPrinter(app(sid + INSTALLED_PRINTER_NAME).ToString())
                ElseIf printerTypeId = 3 Then
                    'use IP-Ethernet printer
                    cpj.ClientPrinter = New NetworkPrinter(app(sid + NET_PRINTER_HOST).ToString(), Integer.Parse(app(sid + NET_PRINTER_PORT).ToString()))
                ElseIf printerTypeId = 4 Then
                    'use Parallel Port printer
                    cpj.ClientPrinter = New ParallelPortPrinter(app(sid + PARALLEL_PORT).ToString())
                ElseIf printerTypeId = 5 Then
                    'use Serial Port printer
                    cpj.ClientPrinter = New SerialPortPrinter(app(sid + SERIAL_PORT).ToString(), Integer.Parse(app(sid + SERIAL_PORT_BAUDS).ToString()), DirectCast([Enum].Parse(GetType(SerialPortParity), app(sid + SERIAL_PORT_PARITY).ToString()), SerialPortParity), DirectCast([Enum].Parse(GetType(SerialPortStopBits), app(sid + SERIAL_PORT_STOP_BITS).ToString()), SerialPortStopBits), Integer.Parse(app(sid + SERIAL_PORT_DATA_BITS).ToString()), DirectCast([Enum].Parse(GetType(SerialPortHandshake), app(sid + SERIAL_PORT_FLOW_CONTROL).ToString()), SerialPortHandshake))
                End If

                'Send ClientPrintJob back to the client

                System.Web.HttpContext.Current.Response.ContentType = "application/octet-stream"
                System.Web.HttpContext.Current.Response.BinaryWrite(cpj.GetContent())
                System.Web.HttpContext.Current.Response.End()

            End If

        End Sub


        Const PRINTER_ID As String = "-PID"
        Const INSTALLED_PRINTER_NAME As String = "-InstalledPrinterName"
        Const NET_PRINTER_HOST As String = "-NetPrinterHost"
        Const NET_PRINTER_PORT As String = "-NetPrinterPort"
        Const PARALLEL_PORT As String = "-ParallelPort"
        Const SERIAL_PORT As String = "-SerialPort"
        Const SERIAL_PORT_BAUDS As String = "-SerialPortBauds"
        Const SERIAL_PORT_DATA_BITS As String = "-SerialPortDataBits"
        Const SERIAL_PORT_STOP_BITS As String = "-SerialPortStopBits"
        Const SERIAL_PORT_PARITY As String = "-SerialPortParity"
        Const SERIAL_PORT_FLOW_CONTROL As String = "-SerialPortFlowControl"
        Const PRINTER_COMMANDS As String = "-PrinterCommands"

        <HttpPost()>
        <AllowAnonymous>
        Public Sub ClientPrinterSettings(sid As String, pid As String, installedPrinterName As String, netPrinterHost As String, netPrinterPort As String, parallelPort As String,
                                             serialPort As String, serialPortBauds As String, serialPortDataBits As String, serialPortStopBits As String, serialPortParity As String, serialPortFlowControl As String,
                                             printerCommands As String)
            Try
                Dim app As HttpApplicationStateBase = HttpContext.Application

                'save settings in the global Application obj

                'save the type of printer selected by the user
                Dim i As Integer = Integer.Parse(pid)
                app(sid + PRINTER_ID) = i

                If i = 2 Then
                    app(sid + INSTALLED_PRINTER_NAME) = installedPrinterName
                ElseIf i = 3 Then
                    app(sid + NET_PRINTER_HOST) = netPrinterHost
                    app(sid + NET_PRINTER_PORT) = netPrinterPort
                ElseIf i = 4 Then
                    app(sid + PARALLEL_PORT) = parallelPort
                ElseIf i = 5 Then
                    app(sid + SERIAL_PORT) = serialPort
                    app(sid + SERIAL_PORT_BAUDS) = serialPortBauds
                    app(sid + SERIAL_PORT_DATA_BITS) = serialPortDataBits
                    app(sid + SERIAL_PORT_FLOW_CONTROL) = serialPortFlowControl
                    app(sid + SERIAL_PORT_PARITY) = serialPortParity
                    app(sid + SERIAL_PORT_STOP_BITS) = serialPortStopBits
                End If

                'save the printer commands specified by the user
                app(sid + PRINTER_COMMANDS) = printerCommands
            Catch ex As Exception
                Throw ex
            End Try
        End Sub
    End Class
End Namespace

File: WCPMVCCS/Controllers/DemoPrintCommandsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

using Neodynamic.SDK.Web;

namespace WCPMVCCS.Controllers
{
    public class DemoPrintCommandsController : Controller
    {
        // GET: DemoPrintCommands
        public ActionResult Index()
        {
            ViewBag.WCPScript = Neodynamic.SDK.Web.WebClientPrint.CreateScript(Url.Action("ProcessRequest", "WebClientPrintAPI", null, HttpContext.Request.Url.Scheme), Url.Action("PrintCommands", "DemoPrintCommands", null, HttpContext.Request.Url.Scheme), HttpContext.Session.SessionID);

            return View();
        }

        //sid: user session id who is requesting a ClientPrintJob
        [AllowAnonymous]
        public void PrintCommands(string sid)
        {
            if (WebClientPrint.ProcessPrintJob(System.Web.HttpContext.Current.Request.Url.Query))
            {
                HttpApplicationStateBase app = HttpContext.Application;

                //Create a ClientPrintJob obj that will be processed at the client side by the WCPP
                ClientPrintJob cpj = new ClientPrintJob();

                //get printer commands for this user id
                object printerCommands = app[sid + PRINTER_COMMANDS];
                if (printerCommands != null)
                {
                    cpj.PrinterCommands = printerCommands.ToString();
                    cpj.FormatHexValues = true;
                }

                //get printer settings for this user id
                int printerTypeId = (int)app[sid + PRINTER_ID];

                if (printerTypeId == 0) //use default printer
                {
                    cpj.ClientPrinter = new DefaultPrinter();
                }
                else if (printerTypeId == 1) //show print dialog
                {
                    cpj.ClientPrinter = new UserSelectedPrinter();
                }
                else if (printerTypeId == 2) //use specified installed printer
                {
                    cpj.ClientPrinter = new InstalledPrinter(app[sid + INSTALLED_PRINTER_NAME].ToString());
                }
                else if (printerTypeId == 3) //use IP-Ethernet printer
                {
                    cpj.ClientPrinter = new NetworkPrinter(app[sid + NET_PRINTER_HOST].ToString(), int.Parse(app[sid + NET_PRINTER_PORT].ToString()));
                }
                else if (printerTypeId == 4) //use Parallel Port printer
                {
                    cpj.ClientPrinter = new ParallelPortPrinter(app[sid + PARALLEL_PORT].ToString());
                }
                else if (printerTypeId == 5) //use Serial Port printer
                {
                    cpj.ClientPrinter = new SerialPortPrinter(app[sid + SERIAL_PORT].ToString(),
                                                              int.Parse(app[sid + SERIAL_PORT_BAUDS].ToString()),
                                                              (SerialPortParity)Enum.Parse(typeof(SerialPortParity), app[sid + SERIAL_PORT_PARITY].ToString()),
                                                              (SerialPortStopBits)Enum.Parse(typeof(SerialPortStopBits), app[sid + SERIAL_PORT_STOP_BITS].ToString()),
                                                              int.Parse(app[sid + SERIAL_PORT_DATA_BITS].ToString()),
                                                              (SerialPortHandshake)Enum.Parse(typeof(SerialPortHandshake), app[sid + SERIAL_PORT_FLOW_CONTROL].ToString()));
                }

                //Send ClientPrintJob back to the client
                System.Web.HttpContext.Current.Response.ContentType = "application/octet-stream";
                System.Web.HttpContext.Current.Response.BinaryWrite(cpj.GetContent());
                System.Web.HttpContext.Current.Response.End();

            }

        }


        const string PRINTER_ID = "-PID";
        const string INSTALLED_PRINTER_NAME = "-InstalledPrinterName";
        const string NET_PRINTER_HOST = "-NetPrinterHost";
        const string NET_PRINTER_PORT = "-NetPrinterPort";
        const string PARALLEL_PORT = "-ParallelPort";
        const string SERIAL_PORT = "-SerialPort";
        const string SERIAL_PORT_BAUDS = "-SerialPortBauds";
        const string SERIAL_PORT_DATA_BITS = "-SerialPortDataBits";
        const string SERIAL_PORT_STOP_BITS = "-SerialPortStopBits";
        const string SERIAL_PORT_PARITY = "-SerialPortParity";
        const string SERIAL_PORT_FLOW_CONTROL = "-SerialPortFlowControl";
        const string PRINTER_COMMANDS = "-PrinterCommands";

        [AllowAnonymous]
        [HttpPost]
        public void ClientPrinterSettings(string sid,
                                             string pid,
                                             string installedPrinterName,
                                             string netPrinterHost,
                                             string netPrinterPort,
                                             string parallelPort,
                                             string serialPort,
                                             string serialPortBauds,
                                             string serialPortDataBits,
                                             string serialPortStopBits,
                                             string serialPortParity,
                                             string serialPortFlowControl,
                                             string printerCommands)
        {
            try
            {
                HttpApplicationStateBase app = HttpContext.Application;

                //save settings in the global Application obj

                //save the type of printer selected by the user
                int i = int.Parse(pid);
                app[sid + PRINTER_ID] = i;

                if (i == 2)
                {
                    app[sid + INSTALLED_PRINTER_NAME] = installedPrinterName;
                }
                else if (i == 3)
                {
                    app[sid + NET_PRINTER_HOST] = netPrinterHost;
                    app[sid + NET_PRINTER_PORT] = netPrinterPort;
                }
                else if (i == 4)
                {
                    app[sid + PARALLEL_PORT] = parallelPort;
                }
                else if (i == 5)
                {
                    app[sid + SERIAL_PORT] = serialPort;
                    app[sid + SERIAL_PORT_BAUDS] = serialPortBauds;
                    app[sid + SERIAL_PORT_DATA_BITS] = serialPortDataBits;
                    app[sid + SERIAL_PORT_FLOW_CONTROL] = serialPortFlowControl;
                    app[sid + SERIAL_PORT_PARITY] = serialPortParity;
                    app[sid + SERIAL_PORT_STOP_BITS] = serialPortStopBits;
                }

                //save the printer commands specified by the user
                app[sid + PRINTER_COMMANDS] = printerCommands;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }


    }
}